docs/dev/writing-code.md

```diff
@@ -192,18 +201,195 @@ TODO: document when to use future vs callback
+But `aws_event_loop_group` has a callback that fires when it finishes shutting down.
+That means C needs to call into Python AFTER the Python `EventLoopGroup` object has been cleaned up.
+
+For C to call into Python, it must reference a Python object
+(the function itself, or an object with a member function).
+This means our binding needs to store a strong reference and
+keep that Python object alive until the callback has fired.
+
+### The Wrong Way to Build it
+
+You MUST NOT create a reference cycle!
+
+You might be tempted to give `event_loop_group_binding` a strong reference to the `EventLoopGroup` instance.
+Then C could simply call a private member function like `EventLoopGroup._on_shutdown_complete()`.
+But this design creates a reference cycle (see image below):
+
+![Diagram of bad binding with reference cycle](./binding-bad-cycle.svg)
+
+### Option 1 - Pass Callbacks to C
+
+Most of our bindings work like this:
+
+![Diagram where C References callbacks](binding-callbacks.svg)
+
+Within `EventLoopGroup.__init__()` a "callable" is defined and passed down to C.
+`event_loop_group_binding` keeps a strong reference to this Python object.
+
+When the final shutdown callback happens in C, the Python callable
+is invoked, and then the reference is released via `Py_DECREF(x)`.
+
+Destruction goes like this:
+*   When the Python code has no references to `elg`, the `EventLoopGroup` instance...
+*   The garbage collector cleans up the `EventLoopGroup`,
+    and the `PyCapsule` referenced by `EventLoopGroup._binding`.
+*   The `PyCapsule` runs its destructor function.
+    *   The destructor function calls `aws_event_loop_group_release(binding->native)`,
+        but doesn't delete the `event_loop_group_binding` struct yet.
+*   The `aws_event_loop_group` won't shutdown until nothing else is referencing it.
+    Even when the final reference is released, it still needs to wait for the threads
+    in its thread-pool to finish their shutdown process.
+*   Finally, shutdown completes and the C callback is invoked.
+    *   The C callback invokes the Python `callable`, then releases it via `Py_DECREF(x)`.
+        *   Now the garbage collector can clean up the `callable` object.
+    *   The C callback finally deletes the `event_loop_group_binding`.
+        This struct only existed to keep two strong references,
+        but now they've both been released.
+
+### Option 2 - Private Core Class
+
+Another option is to build a private `_Core` class that contains anything
+that may need to outlive the main Python object.
+This technique hasn't actually been used, but the author of this doc
+thinks it might be a graceful way to build in the future:
+
+![Diagram with private Core class](./binding-private-core.svg)
 
 # Writing C Code
 
-TODO: Write me.
+TODO:
+Talk about how our tests can and cannot check for leaks.
+Talk about which classes require a `close()` function, and which don't.
 Suggest writing as little C code as possible.
```

Comment:
Super minor nitpick: I would reword this to be something like "as little Python-specific C code as possible", since I think that may better encapsulate the idea (since we generally write most big features in C for cross SDK/CRT use)

Reply:
this is my TODO list man

what I meant here was: write Python instead of C wherever possible because it's like 20 lines of code to read a member variable from C

I went and added a new section earlier in the document about "Don't Add Features"